README.md
Cocos2d Extensions
=================
This repo is a collection of quality 3rd party extensions and additions for the Cocos2D-iPhone Engine.  
Everything that doesn’t modify Cocos2D itself, while bringing new functionality can become a part of Cocos2D-iPhone-Extensions Repo: alternative versions of existing classes, additional categories, new nodes, actions, etc…  

All extensions are well documented,shipped with testcases and are compatible with latest stable Cocos2D-iPhone version.  
With rare exceptions, they are designed for both Mac & iOS.
  
As well as Cocos2D, Cocos2D Extensions are licensed under the MIT License.

Cocos2D Extensions subforum: http://www.cocos2d-iphone.org/forum/forum/17  

Build & Runtime Requirements
====================

  * Mac OS X 10.6, Xcode 3.2.3 (or newer)
  * iOS 3.0 or newer for iOS games
  * Snow Leopard (v10.5) or newer for Mac games

How to get the source
===================== 

```
    git clone git@github.com:cocos2d/cocos2d-iphone-extensions.git
    cd cocos2d-iphone-extensions
	
	# init cocos2d-iphone submodules in order to build & run the tests.
    git submodule update --init
	
	# to get latest source from develop branch, use this command:
	git checkout -t origin/develop
```

Files & Folders
=================
* **cocos2d** - cocos2d-iphone 1.x submodule.
* **cocos2d_2.x** - cocos2d-iphone 2.x submodule.
* **Extensions** - folders with extensions sources, that can be inlcuded in your project.
* **Tests** - sources & resources of Extensions demos.
   * **SharedResources** - resources shared between all tests (icons, fps images, etc...)
   * **SharedSources** - sources shared between all tests (appDelegates, pch's, etc...)
* **cocos2d-extensions-ios.xcodeproj** - XCode Project containing all extensions and their demos/tests for iOS Platform.
* **cocos2d-extensions-mac.xcodeproj** - XCode Project containing all extensions and their demos/tests for Mac OS X Platform.
* **cocos2d_2.x-extensions-ios.xcodeproj** - same as **cocos2d-extensions-ios.xcodeproj**, but uses Cocos2D-iPhone 2.x ( stable tag in "gles20" branch )
* **cocos2d_2.x-extensions-mac.xcodeproj** - same as **cocos2d-extensions-mac.xcodeproj**, but uses Cocos2D-iPhone 2.x ( stable tag in "gles20" branch )

Extensions
=================
 * [iOS/Mac] **CCMenuAdvanced** - CCMenu subclass with additional features: relativeAnchor, more align options, priority property, scrolling with swipe/trackpad/mousewheel
 * [iOS/Mac] **CCMenuItemSpriteIndependent** - CCMenuItemSprite Subclass, that doesnt add normal/selected/disabled images (sprites) as children. It retains them and delegates rect & convertToNodeSpace: methods to normalImage_. So it's possible to use CCSpriteBatchNode & add position sprites of menuItem anyway you want.
 * [iOS/Mac] **CCVideoPlayer** - Simple Video Player for Cocos2D apps.
 * [iOS/Mac] **CCBigImage** - Dynamic Tiled Node for holding Large Images.
 * [iOS/Mac] **CCSlider** - Little Slider Control to allow the user to set the music/sfx/etc level in the range of 0.0f to 1.0f.
 * [iOS/Mac] **CCSendMessages** - CCActionInstant subclass, that is more flexible than other CCActions that run functions. Can be used in many cases as blocks replacement. 
 * [iOS/Mac] **CCScrollLayer** - CCLayer subclass that lets you pass-in an array of layers and it will then create a smooth scroller. Complete with the "snapping" effect.
 * [iOS/Mac] **FilesDownloader** - Downloader for a group of files with shared source path.
 * [iOS/Mac] **TMXGenerator** - Class that generates a single TMX map with multiple layers.
 * [iOS] **CCLayerPanZoom** - CCLayer subclass that can be scrolled and zoomed with one or two fingers (complete with rubber effect, two modes & ability to click through delegate).
 
 Video Overview and more Info can be found on the [Wiki](https://github.com/cocos2d/cocos2d-iphone-extensions/wiki "Wiki")   
 Detailed README for each extension is available in it's folder (i.e. Extensions/CCSlider/README.md).   
 On the GitHub it will be automatically shown under files list in the extension folder.
 
Building & Running Tests
=========================
Agregate target "BuildAllTests" will build all extensions tests - just set it as active target and change only active executable  to choose the test.   
Extension Test Template is used only as a template for new extensions test targets. It should not build, cause there's no ExtensionTest class implementation for this target.   
SYNTHESIZE_EXTENSION_TEST() macro is used (only once in each extension test) to implement ExtensionTest class, that creates scene with default extension test layer.
 
Contributing
================
Looking for Roadmap or TODO's? Check the [issues](https://github.com/cocos2d/cocos2d-iphone-extensions/issues "Issues") page.  
Want to share your own extension for cocos2d? Read this: [Adding-new-Extension](https://github.com/cocos2d/cocos2d-iphone-extensions/wiki/Adding-new-Extension)  
Know something that should be inlcuded in cocos2d-extensions-repo? Got problems and/or found a bug? [Create an Issue](https://github.com/cocos2d/cocos2d-iphone-extensions/issues/new "New Issue")


CHANGELOG

0.2.1 - 22-April-2012
   * Both 1.x/2.x Cocos2D-iPhone Support (Issue #100, Pull Request #99)
      * Added 2.x Xcode projects for Mac/iOS that uses 2.x Engine branch
      * New submodule: Cocos2D-iPhone 2.0-rc1 ( cocos2d_2.x/ folder )
	  * Tested with Cocos2D-iPhone 1.0, 1.1-beta2b, 2.0-rc1
   * CCScrollLayer: compatibility with cocos2d-iphone 1.1-beta2a (Issue #97)
   * CCScrollLayer: automatic menuItems unselection when stealTouches=YES (Sideeffect of fix for #97)
   * CCVideoPlayer: +playMovieWithPath ( Pull Requests #101 & #102 )
   * CCBigImage doesn't change default pixel format anymore
   * Versions of extensions in Copyrights and Javadoc

0.2 - 09-November-2011
   * New Extension: CCLayerPanZoom (iOS Only)
   * New Extension: TMXGenerator.
   * CCSlider: Uses KVO instead of delegate property.
   * ССScrollLayer: Added new property: marginOffset.
   * CCScrollLayer: Restore blending functions properly.
   * CCScrollLayer: minor fixes & additions:
      1. Assign delegate instead of retain.
      2. Avoid double call of scrollLayer:scrolledToPageNumber: (Issue #82).
      3. Call scrollLayer:scrolledToPageNumber: on Mac, using wheel to scroll.
      4. Added ability to change colors of pages indicator (Issues #84, #87).
   * CCMenuAdvanced fixes:   
      1. Bottom to top align bug - menuItem outside of menu bounds (Issue #75).
      2. Inverted X scroll in Mac with wheel.
   * Analyzer issues fixes in CCBigImage & CCMenuAdvanced.
   * Warnings fixes & project modernization.

0.1.3 - 15-August-2011
   * Hotfix: CCVideoPlayer on Mac hangups, when changing fullscreen to/from windowed & skip video after this with any key (not mouse).

0.1.2 - 10-August-2011
   * Critical fix for FilesDownloader to be compatible with Mac OS X File System Guide.
   * Updated Cocos2D-iPhone 1.0.1 Release.
   * Better comments for some extensions - preparing for full Doxygen documentation (JavaDoc Style).
   * FilesDownloader - added support for subpaths in filenames. Locally replicate the folder structure. (Issue #40)
   * CCBigImage - implemented full transform support.
   * CCVideoPlayer: 
      1. ToggleFullscreen support on Mac. (Issue #31)
	  2. Prepare to play to prevent view flicker while movie is preloading.
	  3. Actually hides the controls on pre 3.2 devices.
	  4. Ability to disable video skip by user (key/mouse/touch event).
	  5. Added compatibility with latest FilesDownloader changes (Issue #53 fix) on Mac.
   * A lot for CCScrollLayer: 
      1. Added pagesIndicatorPosition property, moveToPage: & selectPage: methods.
      2. Changed pages numbering to start from zero.
      3. Mac Support & testcase in Mac Project. (Issue #22)
      4. Added Delegate Property.
      5. Added Pages array Property.
	  6. Fixed jumpy behavior when scrolling with two or more fingers.
	  7. Added Dynamic Pages Control.
	  8. Move to nearest page, not just next/prev by swipe/mouseDrag.
	  9. Ability to disable stealing touches (stealTouches property).
	  10. Do not cancel scroll when stealing touch (Issue #38).
   * CCSendMessages: fixed issue #30, added testcase (return nil on initWithTarget: nil).
   * Minor stability, analyzer & documentation issues fixes (CCVideoPLayer, CCBigImage, FilesDownloader).   

0.1.1 - 12-June-2011
   * Edited README: master is a default branch now.
   * Added CHANGELOG.
   * Reduced repo work tree size.

0.1 - 11-June-2011
   * Initial release with following extensions:
      * [iOS/Mac] CCMenuAdvanced
      * [iOS/Mac] CCMenuItemSpriteIndependent 
      * [iOS/Mac] CCVideoPlayer
      * [iOS/Mac] CCBigImage
      * [iOS/Mac] CCSlider
      * [iOS/Mac] CCSendMessages 
      * [iOS]     CCScrollLayer 
      * [iOS/Mac] FilesDownloader


Extensions/TMXGenerator/cencode.c
/*  Copyright (c) 2006-2007, Philip Busch <broesel@studcs.uni-sb.de>
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *   - Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   - Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @file
 * Base64 implementation.
 * @ingroup base64
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cencode.h"

#define XX 100

/** @var base64_list
 *   A 64 character alphabet.
 *
 *   A 64-character subset of International Alphabet IA5, enabling
 *   6 bits to be represented per printable character.  (The proposed
 *   subset of characters is represented identically in IA5 and ASCII.)
 *   The character "=" signifies a special processing function used for
 *   padding within the printable encoding procedure.
 *
 *   \verbatim
 Value Encoding  Value Encoding  Value Encoding  Value Encoding
 0 A            17 R            34 i            51 z
 1 B            18 S            35 j            52 0
 2 C            19 T            36 k            53 1
 3 D            20 U            37 l            54 2
 4 E            21 V            38 m            55 3
 5 F            22 W            39 n            56 4
 6 G            23 X            40 o            57 5
 7 H            24 Y            41 p            58 6
 8 I            25 Z            42 q            59 7
 9 J            26 a            43 r            60 8
 10 K            27 b            44 s            61 9
 11 L            28 c            45 t            62 +
 12 M            29 d            46 u            63 /
 13 N            30 e            47 v
 14 O            31 f            48 w         (pad) =
 15 P            32 g            49 x
 16 Q            33 h            50 y
 \endverbatim
 */
static const char base64_list[] = \
"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static const int base64_index[256] = {
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,62, XX,XX,XX,63,
    52,53,54,55, 56,57,58,59, 60,61,XX,XX, XX,XX,XX,XX,
    XX, 0, 1, 2,  3, 4, 5, 6,  7, 8, 9,10, 11,12,13,14,
    15,16,17,18, 19,20,21,22, 23,24,25,XX, XX,XX,XX,XX,
    XX,26,27,28, 29,30,31,32, 33,34,35,36, 37,38,39,40,
    41,42,43,44, 45,46,47,48, 49,50,51,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
    XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX, XX,XX,XX,XX,
};

/** Encode a minimal memory block. This function encodes a minimal memory area
 *  of three bytes into a printable base64-format sequence of four bytes.
 *  It is mainly used in more convenient functions, see below.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @param len input size in bytes (between 0 and 3)
 * @returns nothing
 *
 * @ingroup base64
 */
void base64_encode_block(unsigned char out[4], const unsigned char in[3], int len)
{
	out[0] = base64_list[ in[0] >> 2 ];
	out[1] = base64_list[ ((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4) ];
	out[2] = (unsigned char) (len > 1 ? base64_list[ ((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6) ] : '=');
	out[3] = (unsigned char) (len > 2 ? base64_list[in[2] & 0x3f] : '=');
}

/** Decode a minimal memory block. This function decodes a minimal memory area
 *  of four bytes into its decoded equivalent. It is mainly used in more
 *  convenient functions, see below.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @returns -1 on error (illegal character) or the number of bytes decoded
 *
 * @ingroup base64
 */
int base64_decode_block(unsigned char out[3], const unsigned char in[4])
{
	int i, numbytes = 3;
	char tmp[4];
	
	for(i = 3; i >= 0; i--) {
		if(in[i] == '=') {
			tmp[i] = 0;
			numbytes = i - 1;
		} else {
			tmp[i] = base64_index[ (unsigned char)in[i] ];
		}
		
		if(tmp[i] == XX)
			return(-1);
	}
	
	out[0] = (unsigned char) (  tmp[0] << 2 | tmp[1] >> 4);
	out[1] = (unsigned char) (  tmp[1] << 4 | tmp[2] >> 2);
	out[2] = (unsigned char) (((tmp[2] << 6) & 0xc0) | tmp[3]);
	
	return(numbytes);
}

/** Compute size of needed storage for encoding. This function computes the
 *  \e exact size of a memory area needed to hold the result of an encoding
 *  operation, not including the terminating null character.
 *
 * @param len input size
 * @returns output size
 *
 * @ingroup base64
 */
size_t base64_encoded_size(size_t len)
{
	return(((len + 2) / 3) * 4);
}

/** Compute size of needed storage for decoding. This function computes the
 *  \e estimated size of a memory area needed to hold the result of a decoding
 *  operation, not including the terminating null character. Note that this
 *  function may return up to two bytes more due to the nature of Base64.
 *
 * @param len input size
 * @returns output size
 *
 * @ingroup base64
 */
size_t base64_decoded_size(size_t len)
{
	return((len / 4) * 3);
}

/** Encode an arbitrary size memory area. This function encodes the first
 *  \c len bytes of the contents of the memory area pointed to by \c in and
 *  stores the result in the memory area pointed to by \c out. The result will
 *  be null-terminated.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @param len input size in bytes
 * @returns nothing
 *
 * @ingroup base64
 */
void base64_encode_binary(char *out, const unsigned char *in, size_t len)
{
	int size;
	size_t i = 0;
	
	while(i < len) {
		size = (len-i < 4) ? len-i : 4;
		
		base64_encode_block((unsigned char *)out, in, size);
		
		out += 4;
		in  += 3;
		i   += 3;
	}
	
	*out = '\0';
}

/** Decode an arbitrary size memory area. This function decodes the
 *  base64-string pointed to by \c in and stores the result in the memory area
 *  pointed to by \c out. The result will \e not be null-terminated.
 *
 * @attention This function can't check if there's enough space at the memory
 *            memory location pointed to by \c out, so be careful.
 *
 * @param out pointer to destination
 * @param in pointer to source
 * @returns -1 on error (illegal character) or the number of bytes decoded
 *
 * @ingroup base64
 */
int base64_decode_binary(unsigned char *out, const char *in)
{
	size_t len = strlen(in), i = 0;
	int numbytes = 0;
	
	while(i < len) {
		if((numbytes += base64_decode_block(out, (unsigned char *)in)) < 0)
			return(-1);
		
		out += 3;
		in  += 4;
		i   += 4;
	}
	
	return(numbytes);
}

/** Encode a string. This is a convenience function. It encodes the first
 *  \c size bytes of the string pointed to by \c in, stores the null-terminated
 *  result in a newly created memory area and returns a pointer to it.
 *
 * @attention After a call to base64_encode(), you have to free() the result
 *  yourself.
 *
 * @param in pointer to string
 * @param size strlen
 * @returns NULL on error (not enough memory) or a pointer to the encoded result
 *
 * @ingroup base64
 */
char *base64_encode(const char *in, size_t size)
{
	char *out;
	size_t outlen;
	
	if(in == NULL)
		return(NULL);
	
	if(size == 0)
		size = strlen(in);
	
	outlen = base64_encoded_size(size);
	
	if((out = (char *)malloc(sizeof(char) * (outlen + 1))) == NULL)
		return(NULL);
	
	base64_encode_binary(out, (unsigned char *)in, size);
	
	return(out);
}

/** Decode a string. This is a convenience function. It decodes the
 *  null-terminated string pointed to by \c in, stores the result in a newly
 *  created memory area and returns a pointer to it. The result will be
 *  null-terminated.
 *
 * @attention After a call to base64_decode(), you have to free() the result
 *  yourself.
 *
 * @param in pointer to string
 * @returns NULL on error (not enough memory) or a pointer to the decoded result
 *
 * @ingroup base64
 */
char *base64_decode(const char *in)
{
	char *out;
	size_t outlen;
	int numbytes;
	
	outlen = base64_decoded_size(strlen(in));
	
	if((out = (char *)malloc(sizeof(char) * (outlen + 1))) == NULL)
		return(NULL);
	
	if((numbytes = base64_decode_binary((unsigned char *)out, in)) < 0) {
		free(out);
		return(NULL);
	}
	
	*(out + numbytes) = '\0';
	
	return(out);
}
